### ISSUE TITLE
re-enable partitioned transfers

### ISSUE THREAD

**Issue by petersilva** [MetPX/sarracenia#396]
v2 implemented transfering of partitioned files.  Partition sizes could be specified, and there was an automatic algorithm to pick partition sizes based on the file size.  This was never used, but it is an ask for very large files, such as multi-gigabyte ones that are written sequentially over a long period.  Examples: tar files built over 10-15 minutes (RCM passes could benefit from this)  or read-outs from DNA sequencers, which are typically multiple gigabytes written out over a one to a few days.

Partition sizes need to be fixed in order to avoid a number of pitfalls.  so we just pick a standard one (for production, we will say.... 50 megabytes) in development, people can pick much smaller ones to make it easier (1k?) and we can leave it user settable, but having it dynamically calculcated based on file size is a problem when the file is being continuously written.  have to think about the partition notation.

**Comment by petersilva** [comment]
first step here is to add to the flow tests a partitiioned file transfer, especially one that v2 passes, and v3 will initially fail.

**Comment by petersilva** [comment]
in v2, the partition files had the entire parts header as part of the name.
```

thefile.5000000,10,1,37.part

```

decoding the name:

* p- partitioned,
* blocksize
* blockcount,
* block number,
* remainder (on the last block)
* '.part' to mark it as a partition.

So each partition would end in *.part* and be passed from one node to the other until it gets reassembled at the far end. The files differ only in *block_number* and are catenated together
in *block_number* order.

In v2, the blocksize was a puzzle. One could specify it as a fixed value, but originally, the ideas was to customize the blocksize based on the file size.

This design had some problems in the following use cases:

Genetic Sequencer Use Case
---------------------------------------

Genetic data, a sequencer is continuously appending to a very large file of a day or so, so idealy, one posts the file multiple times during that period, and blocks progressively transfer over that period.  When blocks are already transferred it should notice, and duplicate suppression should prevent
transferring them again.

* if the file is changing size, then the block count, and remainder will change, so the otherwise potentially identical file parts will not have the same name, creating a need for fuzzy matching.

* if the file is changing size, and the block size might change again changing the part file names, changing the 

So the partition naming scheme is a problem.  the partition file names was the only place this
information was stored.  Once re-assembly occurs on the destination end, in v2, there is no record of the block files. For partial files, need to keep information about the formerly existing partition files.

Distributed Copy Use Case
------------------------------------

In v2, one only stored the partition information in the file name, so if a file was announced at the origin, and then picked up by a transfer cluster, block 1 might be picked up by node 1, block 2 by node 2, etc...   so each node will only have duplicate suppression information on the blocks it transferred itself.  If on the second posting, node2 receives a message about block 1, then it will think it is new.

Things we can do:

* do duplicate suppression at source (before posting for downstream users.) so we only post the blocks that have changed (in the gene sequencer case.) because we record what  the blocks were the first time.
* every partition in the file receives a list of checksums for all files, and thus can "know" whether a block has been posted before.  but it has to know that some other node has successfully transferred the block.
* use shared (e.g. redis ) duplicate suppression so all blocks are in the same suppression cache.



Changes Needed
-----------------------

* make blocksize more predictable & fixed.
* remove the remainder & block_count from the suffix.
* perhaps remove block_size from the suffix? it should be the file size?
* perhaps add xattr of blocks (to retain the remainder & block_count there.)

**Comment by petersilva** [comment]
Questions:
* should we include a complete TofC with each block (checksums and sizes for each block in the file.)
* Should we humanize the blocksize where it is printed (so it  prints more readably as 500m ... )
* Should we add block metadata to the information stored in xattr.

**Comment by petersilva** [comment]
Example:
* sr_post is done of a 1 GB  themainfile, using a 100 MB blocksize.
  * so it posts 10 notifications, each for 1/10th, of themainfile.
* each block is picked up by 1/5 nodes in a PXA (5 independent transfer engines)
  * node 1 gets blocks 0 and 5, node 2 gets 1 and 6, etc...  each one labelled  themainfile.suffix.part
  * what is the part?
* an end consumer re-assembles the blocks into a single file.

an hour later:
* sr_post is repeated.

what happens?

**Comment by petersilva** [comment]
Let's say we store an entire TofC in the extended attributes of the file.
* the tofc contains:
  * the number of blocks and blocksize of the file when it was last posted.
  * the checksums of each block.

Wherease before our method of determining blocksize was to use  blockcount*blocksize + remainder, we could now use the TofC to add up the block sizes.
 
We can calculate the block checksums and and the existing duplicate suppression cache should take care of suppressing them... or just notice that it matches the TofC and discard even earlier?

For the last block it had a different size before.  The old tofC looks something like:

* part 0: bytes 0 to 100m-1, 
* part 1: bytes 100m to 300m-1, .
* .. 
* part9: bytes 900m to 950m

When we post again in v2 we would send a new part9, that contains the 50m already send, and the additional data to round out the block, and then the last block from there.

If we have the existing TofC stored, then perhaps we can post part10 as just 50mb, and it fills in to get to a block boundary, and we can have a "partial_block_count": 1 to make the math still easy.

Perhaps we just add one block that fills in the remainder, and then add a new last block. so the TofC becomes something like:

* part10: bytes  950 to 1g
* part11: bytes 1g to 1.1g+73mb

The new partial_block_count makes it easier to calculate file size...

* size = (block_count-partial_block_count)*size+remainder.

versus what we would otherwise need to do:

* size = sum( blocksize of each block in the TofC )

not sure what's better. This also begs the question, why just fill in the intermediate value? why not just add a full block ? then only the latter method for size calculation works, and one can never predict a byte offset without reading the entire TofC (partial_block_count gives us a shortcut.)

**Comment by petersilva** [comment]
The advantage of the partial block refinement is that we don't resend any data we don't have to.

**Comment by petersilva** [comment]
File Completion
---------------------

Another problem with the v2 partitioning scheme is that it isn't possible to know when a file is complete.  We can receive parts out of order, and we only have the checksum of each part when a part is received, so there is no way to know if all the requisite parts have arrived.   When you receive the last part, there could be some intervening missing block.

If we have a global manifest included with each message, and we include a global manifest as metadata with the file, then once we finish downloading a partition, we can compare the overall manifest from the file on disk and in the message, and if they now match, we have a complete file.  It's now clear and idependent of order sent.

Re-assembly Locking
----------------------------

If we have a global manifest, and there are multiple transfer engines writing to the same file, we need to be sure about serializing access to the manifest (if multiple writers at once each updating different blocks, then processes writing one block could inadvertantly back out of the change to another block,
since the entire manifest is written as one unit.)  Some sort of locking needed to avoid that.

**Comment by petersilva** [comment]
ok so new design is something like

blocks { "... same old fields, plus" "manifest" : { 0: { "size": blksize, "identity": checksum }, 1: {  ... } ... }

* each partition will have a full manifest of all the blocks in the file and their sizes.
* the block will indicate which block (index into the manifest) is the current one.
* not sure if parent "identity" value for the entire file is a good thing or not.
* each downloaded partition knows the block size and which block in the series it is.
* during re-assembly, the checksums of all blocks are known. 
* so drop the "size" from the "blocks" element (replaced by manifest)
* manifest needs to be stored in extended attributes.
* when posting a file for the second time... do it as a delta from the existing extended attributes.
* so the last block will have a smaller size to fill up the last block.
* Have a "partial_blocks" listing.
* don't need "remainder" in the "block" section anymore (replaced by manifest.)
* Don't need block_count any more (just length of manifest.)

putting that in the branch as something to try.

**Comment by petersilva** [comment]
So in the v03_block_partitioned_transfers branch, there is some stuff working:

* when a file is posted with "blocksize > 1" option set, it and the file size is > blocksize.  (blocksize is an existing option from v2... that was used to 
* the post creates a "blocks" field in the message, with a manifest in it.  The manifest lists all the blocks that would be needed for a complete file.
* a post message is produced for each block in the manifest, and each post has a "number" in the "block" set differently.
* when a flow tries to download it ... it only downloads only the portion of the file corresponding to the block.  and it adds a block suffix to the file name.
* block suffix looks like:   §block_0001,20m_§   
   * enclosed by two unicode section header characters, 
   * begin with §block_ 
   * followed by a four digit block number.
   * followed by the standard blocksize of the transfer.
* thereafter the block files transfer as normal files.
* There is a reassemble flow callback, that can be added:  "callback_prepend reassemble"
* the callback is prepended so it is done early.  
   * the main entry point is after_work doing the the reassembly.
   * other after_work plugins will only need to operate on files after they are reassembled.
* when that callback is loaded, it will notice block suffixed files
   * It will determine the name of the final file (removing the block suffix) calling that the root_file.
   * It will put the block in the right part of the root_file (seek to offset as needed.)
   * after copying the data into the root_file, it will unlink the block file.
   * It will remove messages concerning block files from the flow, 
   * once the root_file is complete, it will produce a message for downstream consumption representing the entire file.

**Comment by petersilva** [comment]
This is the branch for this work: https://github.com/MetPX/sarracenia/tree/v03_block_partitioned_transfers

**Comment by petersilva** [comment]
branch status:
* have a new flow test directory... *partitioned_flow* in the insects test suite.  It started as a copy of the *static_flow*.
* uses in-place posting (post/t_dd* has *blocksize 200* for 200 byte blocks.)
* The subscriber to the initial posts is sarra/download_f20 component, which does the download, partitioning as it goes (output is a series of blocks files for files that are big enough.)
* so the initial download tree (~/sarra_devdocroot/YYYYMMDD ) is quite different from the static source, having hundreds of block partition files.
* a second subscriber of the sarra (subscribe/amqp_f30) runs the reasssembly, building the original files back from the partitions.
* all static/flakey/dyn tests (that doe not activate partitioning) pass.
* on the partitioning tests, 27/32 tests pass.
* there is some corruption... in the reassembly... a handful of files... 
* after the reassembly, the files are posted by watch (normally) and the rest is the same as static_flow.
* I'm not sure if watch is posting some partially re-assembled files.

currently diagnosing the corruption.

Use cases perhaps not tested yet:
* re-posting a growing file.
* inflight != NONE.... all inflight is currently with NONE

## Merge? or Long running branch?

This branch started a few weeks ago, has merged the 3.0.44 release back into it.
Puzzling over "commit early/commit often" this this branch that does no harm be merged soon? and then another branch established to continue?   partitioning stuff is labelled as experimental and all other tests still work.  I worry about long running parallel branches.

**Comment by petersilva** [comment]
OK, so the problem causing missing files is that there are limits on the size of extended attributes that can be written.  This raises a number of issues:
* current code, treats extended attributes as a "bonus" extra metadata used to accellerate things.  when missing, it just means re-calculating things like checksums. so when they fail, it is just ignored.
* but for block transfers, it is the only source of block re-assembly information.  so writing the information to disk is essential and when it fails, need to report back to the main code.
* the problem we see is that, for large files, the block lists are too big to write as a single attribute.

so... things to think about:
* All the re-assembly stuff is actually correct (works fine for smaller files.) the problems are with extended attributes... so in that sense... merging might make sense.

* likely want to re-consider how the block ranges are written, instead of a dumb python list, perhaps a group of ranges, and it will get much shorter then.  That would involve improvements to all the code that writes block lists.

* probably want to re-factor to be able to optionally notice when setxattr fails.

**Comment by petersilva** [comment]
After exploring how to deal with really big extended attributes in #779, the conclusion is that it is a bad idea.  So thinking about revising the design as follows:

* when you post a multi-block file (in place) you also create a block_manifest file (ending in §block_manifest_200b_§ ) put the manifest data in that file.
* when a downloader downloads from an "inplace" file, it should look if there is a manifest file present.  If the file mtime in the message is newer than the manifest mtime, then it should replace the manifest??? thinking... update the manifest?
* when a downloader downloads a §block_ file, if the manifest is not in the message, then it looks for one in the same directory on the server.  If it doesn't find it, debug message (... might be able to get the manifest from somewhere else.)
* while downloading the §block... check and update the §block_manifest on local host.
* All access to §block_manifest needs to be mediated by locking. 
* when gathering a §block_ file from disk, look for the corresponding §block_manifest file, and add the data in there to the message.
* when reassembling the file, store the necessary information in the §block_reassembly  (another file?) §block_manifest is what is on disk at any given time, §block_reassembly is what we are still waiting for.  Once the reassembly is complete, that file disappears, but the §block_manifest stays.
* when sftp'ing a file, can put the §block_manifest and $block_reassembly files to remote, until reassembly is complete. (sftp reassembly on push is a stretch goal... later...)

After we have added the management of the manifest file, we may be able to remove the manifest data from the message itself... tbd.

**Comment by petersilva** [comment]
Second branch v03_issue396_2_manifest_files seems to work mostly.  have 25/32 tests passing in partitioned_flow.  Implementation:

* blockmanifest class, writes the manifests to a metadata file.
* flowcb/block_reassembly.py (renamed flowcb/reassemble.py)
* examples/block_reassembly.inc
* some changes in flowcb/gather/file.py, and flow/__init__.py

 looking at tree differences at end of test. there are a handful of differences in the trees.  needs investigation.

**Comment by petersilva** [comment]
This was all merged... but it isn't clear if it is working yet.

**Comment by petersilva** [comment]
There is a *partitioned_flow* in the Flow_tests/AMQP that every merge to the development branch and all current releases.  So I think this works.
